### DISCUSSION TITLE
organisation/style/docs  comments based on thoughts from ls_dev branch

### DISCUSSION THREAD

**Discussion by L-Siemons** [L-Siemons/SpinSimulation#8] [General]
in the ls_dev branch I made some changes and I have the following questions which I guess we should address. 

1) The spin indexes are integers and according to PEP they should be given small letters as names but in NMR we like to use I/S as capitals to indicate spins. Which convention should we use ? I am tempted to lean towards PEP as it tells us about what types the variables are. 

2) Currently we have a mix of doc string types. We should just pick on. At the moment it's mostly the numpy docs style. Can we use this one? 

example:
```
Takes the Hamiltonian and applies it to the density matrix System.rho

Parameters
----------
hamiltonian : numpy.ndarray
    The Hamiltonian you want to evolve
time : float
    time in seconds
```

3) Do we want to make a seperate submodule called math_funcs for useful maths functions like sparse_dot. 
It seems odd to have this in the spin submodule because if its needed else where then it will be in the
wrong place as its not really related to spins. Otherwise we could make it a method in system. 

4) Do we want to commit to using longer names? In general I find that for medium length names I prefer to write them out. I 
overall find it easier to read  
```
- operator 
- amplitude
- propagator 
```
over 
```
- ops
- ampl
- prop
```

5) Do we want to keep the following functions in spin.System ?
system.apply_rotation()
system.apply_hamiltonian()
system.apply_hamiltonian_to_self()
system.get_trace()
system.scalar_coupling_hamil()
system.calc_fid()

Some of these act on system.rho and somehow I felt that we were thinking of moving away from this type of structure to one where rho was kept outside of this class and it just provided functionality for making operators and hamiltonians.

**Comment by phagost** [comment]
1. I am for PEP
2. I introduced the mix of docstrings, and I agree to use numpy docstrings which are mostly in the code right now. I will clean this up in future commits.
3. We can reorganize the method to be in a separate file or as a class method, I don't have a particular preference.
4. I prefer `op, ampl, prop`, because they will be used quite often, it is shorter to write and I don't have a problem understanding these functions in the code. Moreover, in the future, we will have commands like `double_commutator_superoperator` which is indeed long to write, but if one has established short notation like `operator` as `op` and `superoperator` as `s_op` or `sop`, the length of this function's names will be significantly reduced
5. Maybe I would leave some of these functions, for example plottting a spectra from a given density matrix. But all functionality of these methods should be in a different folder as functions and not as methods of the class.

**Comment by L-Siemons** [comment]
1) So with regards to PEP8 one thing that we could try to aspire to would be to lint/format the code with either Pylint or Black before merging it into main. This will be a bit painful the first time but should give the whole repository a consistent style that reflects the object types and so on. Rajka did this for what is currently on the main branch. It could also be cool to have this in the git workflow. If you are happy with it I can look into it. 

3) Ok let's use a `math_funcs` module. I have done this for other repositories and like it - but I might be biased. Again I can do this. 

4) On the point of naming. I much prefer them to be longer and I know we type them alot but we can also auto tab complete them. That being said, while I would still go with `double_commutator_superoperator` as a name I can see the point that it starts to get a little long. So what I want to say is that short names are fine (sorry for the rambling).

Maybe somewhere it would be nice to have a little dictionary to refer to so that we have recurring abbreviations listed somewhere. It could even be in the readme for now eg:

```
Abbreviations:
--------------
op - operator
sop - super operator 
```

This could make things clearer and also encourage us to always use the same abbreviations. It would be annoying to sometimes see `sop` and then in another function `s_op`. 

5) with regards to the functions. I had a look at what you wrote and now I kinda like this class that just does operator things. I guess down the line if we want we can make a class that has some higher level functionality and combines it. But for now we could move those functions/methods.
